lib/HeatpumpIR/HeatpumpIRFactory.cpp

```diff
@@ -0,0 +1,77 @@
+    return new FujitsuHeatpumpIR();
+  } else if (strcmp_P(modelName, PSTR("gree")) == 0) {
+    return new GreeGenericHeatpumpIR();
+  } else if (strcmp_P(modelName, PSTR("greeyaa")) == 0) {
+    return new GreeYAAHeatpumpIR();
+  } else if (strcmp_P(modelName, PSTR("greeyan")) == 0) {
+    return new GreeYANHeatpumpIR();
+  } else if (strcmp_P(modelName, PSTR("hisense_aud")) == 0) {
+    return new HisenseHeatpumpIR();
+  } else if (strcmp_P(modelName, PSTR("hitachi")) == 0) {
+    return new HitachiHeatpumpIR();
+  } else if (strcmp_P(modelName, PSTR("hyundai")) == 0) {
+    return new HyundaiHeatpumpIR();
+  } else if (strcmp_P(modelName, PSTR("ivt")) == 0) {
+    return new IVTHeatpumpIR();
+  } else if (strcmp_P(modelName, PSTR("midea")) == 0) {
+    return new MideaHeatpumpIR();
+  } else if (strcmp_P(modelName, PSTR("mitsubishi_fa")) == 0) {
+    return new MitsubishiFAHeatpumpIR();
+  } else if (strcmp_P(modelName, PSTR("mitsubishi_fd")) == 0) {
+    return new MitsubishiFDHeatpumpIR();
+  } else if (strcmp_P(modelName, PSTR("mitsubishi_fe")) == 0) {
+    return new MitsubishiFEHeatpumpIR();
+  } else if (strcmp_P(modelName, PSTR("mitsubishi_heavy_fdtc")) == 0) {
+    return new MitsubishiHeavyFDTCHeatpumpIR();
+  } else if (strcmp_P(modelName, PSTR("mitsubishi_heavy_zj")) == 0) {
+    return new MitsubishiHeavyZJHeatpumpIR();
+  } else if (strcmp_P(modelName, PSTR("mitsubishi_heavy_zm")) == 0) {
+    return new MitsubishiHeavyZMHeatpumpIR();
+  } else if (strcmp_P(modelName, PSTR("mitsubishi_kj")) == 0) {
+    return new MitsubishiKJHeatpumpIR();
+  } else if (strcmp_P(modelName, PSTR("mitsubishi_msc")) == 0) {
+    return new MitsubishiMSCHeatpumpIR();
+  } else if (strcmp_P(modelName, PSTR("mitsubishi_msy")) == 0) {
+    return new MitsubishiMSYHeatpumpIR();
+  } else if (strcmp_P(modelName, PSTR("mitsubishi_sez")) == 0) {
+    return new MitsubishiSEZKDXXHeatpumpIR();
+  } else if (strcmp_P(modelName, PSTR("panasonic_ckp")) == 0) {
+    return new PanasonicCKPHeatpumpIR();
+  } else if (strcmp_P(modelName, PSTR("panasonic_dke")) == 0) {
+    return new PanasonicDKEHeatpumpIR();
+  } else if (strcmp_P(modelName, PSTR("panasonic_jke")) == 0) {
+    return new PanasonicJKEHeatpumpIR();
+  } else if (strcmp_P(modelName, PSTR("panasonic_lke")) == 0) {
+    return new PanasonicLKEHeatpumpIR();
+  } else if (strcmp_P(modelName, PSTR("panasonic_nke")) == 0) {
+    return new PanasonicNKEHeatpumpIR();
+  } else if (strcmp_P(modelName, PSTR("samsung_aqv")) == 0) {
+    return new SamsungAQVHeatpumpIR();
+  } else if (strcmp_P(modelName, PSTR("samsung_fjm")) == 0) {
+    return new SamsungFJMHeatpumpIR();
+  } else if (strcmp_P(modelName, PSTR("sharp")) == 0) {
+    return new SharpHeatpumpIR();
+  } else if (strcmp_P(modelName, PSTR("toshiba_daiseikai")) == 0) {
+    return new ToshibaDaiseikaiHeatpumpIR();
+  } else if (strcmp_P(modelName, PSTR("toshiba")) == 0) {
+    return new ToshibaHeatpumpIR();
+  }
```

Comment:
As you're calling this factory class quite often, maybe use a switch statement on the first letter of the model name instead of chaining all options in `else if` checks.
Then you still may need to have a few `else if` checks to distinguish between the various Panasonic or Mitsubishi models, but at least you don't have to test all other brands first.

Now a "Sharp" or "Toshiba" user will need a lot more checks than needed.

Reply:
This shouldn't really be an issue. Typically the air conditioner / heat pump state is not changed often, I would guess not more than max. a couple of times per day.